### PR TITLE
Request and Response Objects refactor

### PR DESCRIPTION
**Pull request by kfrankc**
### API Changes

Created Request and Response Objects for storing data from request and response buffers, which we can easily access.

**Request Object**

Getter Functions:

* `raw_request()` - returns the raw request buffer string
* `method()` - returns the method string
* `uri()` - returns the URI string
* `version()` - returns the version string
* `headers()` - returns the headers in formatted HTTP string
* `body()` - returns the body string

**Response Object**

Setter Functions

* `SetStatus(response_code)` - set status code 
* `AddHeader(header_name, header_value)` - add header name and header value as a pair in the vector of headers
* `SetBody(body)` - set body
* `toString()` - convert response object's private variables to a single formatted HTTP string to return

### CODE REVIEW

**Comment by KyCodeHuynh** [comment]
@kfrankc: Could we add a description to this PR? 😺

File: lightning_server.cc [hunk 1]
```diff
@@ -54,6 +56,20 @@ void LightningServer::start() {
         continue;
     }
 
+    // DEBUG Testing Below; Can be removed later
```
Comment (KyCodeHuynh) [inline lightning_server.cc]: Could we remove this commented-out code? You could save it in a stash or local branch to preserve it, but this lets us keep the codebase clean.

File: request.cc [hunk 2]
```diff
@@ -0,0 +1,120 @@
+#include "request.h"
+
+#include <boost/tokenizer.hpp>
+#include <boost/algorithm/string.hpp>
+
+using Headers = std::vector<std::pair<std::string, std::string>>;
```
Comment (KyCodeHuynh) [inline request.cc]: Is this using-declaration needed even if the header already defined it?

File: request.cc [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+// Helper function to tokenize a string
+boost::tokenizer<boost::char_separator<char>> 
+token_generator(const std::string s, const char* sep) {
```
Comment (KyCodeHuynh) [inline request.cc]: Could we rename this helper to follow our method naming convention of `camelCase()`?

File: request.cc [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  int pos = raw_request.find_first_of('\r\n');
+
+  std::string first_line = raw_request.substr(0, pos);
+
+  std::string rest_lines = raw_request.substr(pos+1);
+  std::cout << "first_line: " + first_line << std::endl;
+  std::cout << "rest_lines: " + rest_lines << std::endl;
+
+  // Replace '\r\n' with space for tokenizing later
+  // boost::replace_all(first_line, "\r\n", "");
```
Comment (KyCodeHuynh) [inline request.cc]: Could we remove this commented-out code? You could save it in a stash or local branch to preserve it, but this lets us keep the codebase clean.

File: request.cc [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  // rest of lines depends on ": " and "\r\n" as separators for key/value
+  boost::tokenizer<boost::char_separator<char>> tokens1 = token_generator(
+                                                          first_line, " ");
+  boost::tokenizer<boost::char_separator<char>> tokens2 = token_generator(
+                                                          rest_lines, "~~~");
+
+  // Tokenize first_line
+  // We assume the 2nd token is the uri: GET /path/to/resource.txt
+  int i = 0;
+  auto curToken1 = tokens1.begin();
```
Comment (KyCodeHuynh) [inline request.cc]: For distinguishability, you may want to consider renaming this to `cur_token_first_line`, which would also match our `snake_case` variable convention.

File: request.cc [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  for (;;) {
+    // std::cout << "token: " + *curToken1 << std::endl;
+    if (curToken1 == tokens1.end()) {
+      // TODO: tokens is too short
+      std::cout << "DEBUG: tokens in RequestRouter is too short" << std::endl;
+      break;
```
Comment (KyCodeHuynh) [inline request.cc]: It currently looks like we always this `break`, even if we have the right number of tokens. Do we need a break after the last expected token is processed?

File: request.cc [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+  // Tokenize rest_lines
+  i = 0;
+  std::string headerKey = "";
+  auto curToken2 = tokens2.begin();
+  for (;;) {
+    if (curToken2 == tokens2.end()) {
+      // TODO: tokens is too short
+      std::cout << "DEBUG: tokens in RequestRouter is too short" << std::endl;
+      break;
```
Comment (KyCodeHuynh) [inline request.cc]: [Same as above](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/49/files#r102286589)

File: request.cc [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+      req->version_ = *curToken1;
+    }
+    i++;
+    curToken1++;
+  }
+
+  // Tokenize rest_lines
+  i = 0;
+  std::string headerKey = "";
+  auto curToken2 = tokens2.begin();
```
Comment (KyCodeHuynh) [inline request.cc]: [Similar to above](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/49/files#r102286150). We would also want `header_key` as well.

File: response.h [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+#include <memory>
+#include <string>
+#include <boost/tokenizer.hpp>
+#include <map>
+
+// Represents an HTTP response.
+//
+// Usage:
+//   Response r;
+//   r.SetStatus(RESPONSE_200);
```
Comment (KyCodeHuynh) [inline response.h]: Does our [enum](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/49/files#diff-4ce81ec3dde55b3a00d9839d7bf10049R24) below need to be updated to match this enum value example? How strict is the API requirement on that? 

UPDATE: It doesn't look [any other team](https://github.com/search?q=org%3AUCLA-CS130+RESPONSE_200&type=Code) actually used `RESPONSE_200`. Even the example enum given has `OK = 200`. This is an inconsistency in the spec, and we should just stick to `OK` for brevity.

File: response.cc [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+std::string Response::ToString() {
+  std::string response_str = "";
+  for (size_t i = 0; i < response_header_.size(); i++) {
+    response_str = response_str + response_header_[i].first + ": "
+                   + response_header_[i].second + "\r\n";
+  }
+  return first_line_ + "\r\n" + response_str + "\r\n" + 
+         response_body_ + "\r\n";
+}
```
Comment (KyCodeHuynh) [inline response.cc]: Following [UNIX conventions](https://robots.thoughtbot.com/no-newline-at-end-of-file), can we add a newline at the end of this file?

File: request.cc [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+  int pos = raw_request.find_first_of('\r\n');
+
+  std::string first_line = raw_request.substr(0, pos);
+
+  std::string rest_lines = raw_request.substr(pos+1);
+  std::cout << "first_line: " + first_line << std::endl;
+  std::cout << "rest_lines: " + rest_lines << std::endl;
+
+  // Replace '\r\n' with space for tokenizing later
```
Comment (apfong) [inline request.cc]: This comment should be updated to say replace `\r\n` with `~~~` or update lines 34 and 35 to reflect the comment.

File: response.cc [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+void Response::SetBody(const std::string& body) {
+  response_body_ = body;
+}
+
+std::string Response::ToString() {
+  std::string response_str = "";
+  for (size_t i = 0; i < response_header_.size(); i++) {
+    response_str = response_str + response_header_[i].first + ": "
```
Comment (KyCodeHuynh) [inline response.cc]: Very minor, but would this be more accurately labeled `response_headers_str`? Then the last line becomes: `return first_line_ + "\r\n" + response_headers_str + "\r\n" +  response_body_ + "\r\n";`

File: request.cc [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+  // boost::replace_all(first_line, "\r\n", "");
+  boost::replace_all(rest_lines, "\r\n", "~~~");
+  boost::replace_all(rest_lines, ": ", "~~~");
+  // std::cout << rest_lines << std::endl;
+
+  // Generate tokens
+  // first line depends on " " as separatorrs
+  // rest of lines depends on ": " and "\r\n" as separators for key/value
```
Comment (apfong) [inline request.cc]: This comment should be updated to say rest of lines depends on `"~~~"`

File: request.cc (hunk 6 above)
Comment (apfong) [inline request.cc]: One thing we could do to handle a malformed request is return a `nullptr`

File: request.cc [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    else if (i == 1) {
+      // set uri_ as second token
+      req->uri_ = *curToken1;
+    }
+    else if (i == 2) {
+      // set version_ as third token
+      req->version_ = *curToken1;
+    }
+    i++;
+    curToken1++;
```
Comment (apfong) [inline request.cc]: If we have more than 3 tokens in the first line do we have a bad request?

File: request.cc (hunk 2 above)
Comment (kfrankc) [inline request.cc]: Thanks for the catch! Not needed, but I do need to replace `Headers` with `Request::Headers` now

File: response.h (hunk 9 above)
Comment (kfrankc) [inline response.h]: Thanks for clarifying on that!

File: request.cc (hunk 6 above)
Comment (kfrankc) [inline request.cc]: I changed the for loop from an infinite one to a one that uses tokens count; thanks for the catch! This improves style too!

File: response.h [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+//   r.SetBody(...);
+//   return r.ToString();
+//
+// Constructed by the RequestHandler, 
+// after which the server should call ToString
+// to serialize.
+class Response {
+ public:
+  using Headers = std::vector<std::pair<std::string, std::string>>;
```
Comment (kfrankc) [inline response.h]: Thanks for the catch Alex! No it does not; it can be made private.

File: request.cc (hunk 14 above)
Comment (kfrankc) [inline request.cc]: Yes, that should be handled properly.

**Pull request merged**